Describe B&A payload size optimization options (#1183)
WICG/turtledove: FLEDGE_browser_bidding_and_auction_API.md

  // ‘seller’ works the same as for runAdAuction.
  'seller': 'https://www.example-ssp.com',
  // 'coordinatorOrigin' of the TEE coordinator, defaults to
  //'https://publickeyservice.pa.gcp.privacysandboxservices.com/' (for now). Used to
  // select which coordinator to use for fetching the key used to encrypt the request
  // blob. Will eventually be required.
  'coordinatorOrigin':
    'https://publickeyservice.pa.gcp.privacysandboxservices.com/',
  // 'requestSize' the affects the size of the returned request (optional).
  'requestSize': 51200,
  // 'perBuyerConfig' specifies per-buyer options for size optimizations when
  // constructing the blob (optional).

>   'perBuyerConfig': {'https://buyer1.origin.example.com': {

Reviewer: ```suggestion
  'perBuyerConfig': {
    'https://buyer1.origin.example.com': {
```

Author: Done

      // 'targetSize' specifies the size of the blob to devote to this buyer
      // (optional).
      "targetSize": 8192,
    },
    'https://buyer2.origin.example.com': {}
  }
});
```

The `seller` field will be checked to ensure it matches the `seller` specified

> in the AuctionConfig passed to `runAdAuction()` with the response. The

Reviewer: ```suggestion
in the auction configuration passed to `runAdAuction()` with the response. The
```

Author: Done

Note that since the `serverResponse` field in the config is a promise it is
possible to run the on-device auction in parallel with the B&A auction.

## Request Size and Configuration

### Request Size Controls
The `requestSize` field provided to `navigator.getInterestGroupAdAuctionData()`
can be used to specify the maximum size of the returned request. If the
`perBuyerConfig` field is not present (or is empty) then the `requestSize` field
sets the maximum size for the returned encrypted blob. If the blob fits into a

> size bucket smaller than `requestSize` then that size will be used instead.

Reviewer: Isn't this only true when there is no `perBuyerConfig`?  If so, we should mention it here, otherwise the next sentence contradicts it.

Author: Removed the redundant sentence and moved the "If the..." to be with the correct sentence.

If the `perBuyerConfig` field is specified and non-empty, the returned encrypted

> blob will be exactly `requestSize` bytes long unless there was an error. If an error

Reviewer 2: As a seller concerned about request size, `perBuyerConfig` returning the exact `requestSize` is quite suboptimal. While it's highly desirable to fairly fit multiple buyers on the request, we would prefer to take the min(actual_size, request_size) instead.

Could we satisfy the privacy constrain in a way that better optimizes for streamlined request payload size?

Author: I agree that returning the exact request size is unfortunate, but returning even a bucketed size leaks significantly more data than we are comfortable with.

One alternative that we considered was allowing the config to be specified out of band  per seller, and only allowed to change relatively slowly (preventing the leak from calling the API multiple times with the slightly different parameters). Unfortunately this can be worked around using multiple sellers that have configs that are slightly different. The sellers don't even have to be cooperating because the attack can be completely client-side.

occured then the returned blob will be 0 size and the function may throw an error.

### Buyer Controls
If the `perBuyerConfig` field is specified then the blob will only include the
interest groups for the buyers specifically listed. If `requestSize` is not
specified then the sum of the `targetSize` of each buyer will be used for the
`requestSize`. Interest groups for each buyer will be added to the request -- in
decreasing order by the interest group's `priority` field -- until the next
interest group does not fit.

> Space can be allocated between different buyers in several different modes:

Reviewer: How does the browser know which mode to use?

Author: It's based on which features of the perBuyerConfig are used across all buyers. The details of when each mode is used are specified with the mode description, below. I clarified this sentence a bit though.

1. Equally - Space is divided equally between buyers when `targetSize` is not specified.

2. Proportionally - Space is assigned to buyers using `targetSize` as the
   weight. So a buyer with a `targetSize` of 2 will can use twice as much space
   as a buyer with a `targetSize` of 1. If all buyers have a `targetSize`
   specified then proportional mode is used.

3. Fixed/Mixed - Space is first allocated to buyers with `targetSize` specified.
   They get up to `targetSize` bytes. The remaining space is divided equally
   between the buyers that did not have `targetSize` specified.

> The browser will process first process buyers with `targetSize` specified before

Reviewer: ```suggestion
The browser will first process buyers with `targetSize` specified before
```

Author: Done
